cpp/util/fake_etcd.cc

```diff
@@ -118,39 +156,86 @@ void FakeEtcdClient::Get(const Request& req, GetResponse* resp, Task* task) {
+  if (it == entries_.end()) {
+    task->Return(Status(util::error::NOT_FOUND, "not found"));
+    return;
+  }
+  resp->node = it->second;
+  ++it;
+
+  vector<Node*> parent_nodes{&resp->node};
+  while (!parent_nodes.empty() && it != entries_.end()) {
+    const string key_prefix(parent_nodes.back()->key_ + "/");
+    if (it->first.compare(0, key_prefix.size(), key_prefix) > 0) {
+      parent_nodes.pop_back();
+      continue;
     }
-    resp->node = Node(1, 1, req.key, true, "", move(nodes), false);
-  } else {
-    const map<string, Node>::const_iterator it(entries_.find(req.key));
-    if (it == entries_.end()) {
-      task->Return(Status(util::error::NOT_FOUND, "not found"));
-      return;
+
+    if (req.recursive) {
+      parent_nodes.back()->nodes_.emplace_back(it->second);
+
+      if (it->second.is_dir_) {
+        // The node we just added is now the current parent node.
+        parent_nodes.push_back(&parent_nodes.back()->nodes_.back());
+      }
+    } else {
+      if (it->first.find_first_of('/', key_prefix.size()) == string::npos) {
+        resp->node.nodes_.emplace_back(it->second);
+      }
     }
-    resp->node = it->second;
+
+    ++it;
   }
   task->Return();
 }
 
 
-void FakeEtcdClient::InternalPut(const string& key, const string& value,
+void FakeEtcdClient::InternalPut(const string& rawkey, const string& value,
                                  const system_clock::time_point& expires,
                                  bool create, int64_t prev_index,
                                  Response* resp, Task* task) {
-  CHECK_GT(key.size(), 0);
-  CHECK_EQ(key.front(), '/');
+  const string key(NormalizeKey(rawkey));
   CHECK_NE(key.back(), '/');
   CHECK(!create || prev_index <= 0);
 
+  vector<string> parents;
+  for (string::size_type offset = 0; offset < key.size();) {
+    const string::size_type next_slash(key.find_first_of('/', offset + 1));
+    if (next_slash == string::npos) {
+      break;
+    }
+
+    parents.emplace_back(key.substr(offset + 1, next_slash - offset - 1));
+    offset = next_slash;
+  }
+
   *resp = EtcdClient::Response();
   unique_lock<mutex> lock(mutex_);
   PurgeExpiredEntriesWithLock(lock);
   const int64_t new_index(index_ + 1);
+
+  // If we're creating, make sure all the parent entries exist and are
+  // directories, creating them as needed.
+  if (create) {
+    string parent_key;
+    for (const string& path : parents) {
+      parent_key.append("/" + path);
+      // Either this inserts a directory, or it fails, and it gets us
+      // the entry. If we insert a directory, then success is
+      // guaranteed below, so we're sure to update index_
+      // appropriately.
+      if (!entries_.insert(make_pair(parent_key,
+                                     Node(new_index, new_index, parent_key,
+                                          true, "", {}, false)))
+               .first->second.is_dir_) {
+        task->Return(Status(util::error::ABORTED, "Not a directory"));
```

Comment:
I was "cloning" the behaviour of ```EtcdClient``` which uses the message from the server (which is exactly this). It doesn't really matter much, and I suppose code that would depend on this would be wrong to do so anyway... Keep it? Change it?